### PR TITLE
Istio helm install and upgrade

### PR DESCRIPTION
**Pull request by pfmackin**
This PR changes Istio to use Helm install and implements the Istio upgrade.  Helm install is needed to use Helm upgrade, otherwise Helm doesn't know that Istio is installed

1) Change 1-install-istio.sh to do a Helm install rather than generate manifests then kubectl apply  
2) Change 1-install-istio.sh to use the Istio component override yaml. 
3) Change the docker image build to replace the Istio charts usage kubectl image with istio_kubectl. 
4) Add the component/istio-values.yaml Helm override file. 
5) Add the Istio upgrade component with a Helm prehook to delete the postinstall job

### CODE REVIEW

File: operator/internal/component/istio.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+// PreUpgrade deletes the Istio Helm install job that may have been left over
+// from the previous install.  This must be done or the helm install will fail because the Istio
+// Helm post-hook won't be able to create the job
+func PreUpgrade(client clipkg.Client, _ string, namespace string, chartDir string) error {
+	var log = ctrl.Log.WithName("upgrade")
+
+	chart, err := helm.GetChartInfo(chartDir)
+	if err != nil {
+		log.Error(err, fmt.Sprintf("Unable to get the chart version from %s", chartDir))
```
Comment (bobdonat) [inline operator/internal/component/istio.go:25]: nit. should the message say "Unable to get chart info from %s

File: operator/internal/component/registry.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+			valuesFile:              filepath.Join(componentDir, "keycloak-values.yaml"),
+		},
+		helmComponent{
+			releaseName:             "rancher",
+			chartDir:                filepath.Join(config.Get().ThirdpartyChartsDir, "rancher"),
+			chartNamespace:          "cattle-system",
+			ignoreNamespaceOverride: true,
+			valuesFile:              filepath.Join(componentDir, "rancher-values.yaml"),
+		},
+		helmComponent{
```
Comment (bobdonat) [inline operator/internal/component/registry.go:47]: should these components be in the same order as the install order

File: operator/scripts/install/1-install-istio.sh [hunk 3]
```diff
@@ -68,65 +68,43 @@ function install_istio()
...
 
-    log "Create helm template for installing istio proper"
-    helm template istio ${ISTIO_CHART_DIR} \
+    log "installing istio"
+    helm install istio ${ISTIO_CHART_DIR} \
         --namespace istio-system \
-        --set global.hub=$GLOBAL_HUB_REPO \
-        --set global.tag=$ISTIO_VERSION \
+        -f $SCRIPT_DIR/components/istio-values.yaml \
         --set gateways.istio-ingressgateway.type="${INGRESS_TYPE}" \
```
Comment (bobdonat) [inline operator/scripts/install/1-install-istio.sh:103]: this can be included in EXTRA_HELM_ARGUMENTS with the other --set commands

**Review by bobdonat — COMMENTED** [review]
A few small comments

File: operator/controllers/upgrade_reconciler.go [hunk 4]
```diff
@@ -73,16 +75,22 @@ func isLastCondition(st installv1alpha1.VerrazzanoStatus, conditionType installv
...
+// generation, meaning the last time the CR spec was modified by the user
+func upgradeFailureCount(st installv1alpha1.VerrazzanoStatus, generation int64) int {
 	var c int
 	for _, cond := range st.Conditions {
 		switch cond.Type {
-		case installv1alpha1.UpgradeComplete:
-			c = 0
 		case installv1alpha1.UpgradeFailed:
-			c++
+			if strings.Contains(cond.Message, fmtGeneration(generation)) {
```
Comment (bobdonat) [inline operator/controllers/upgrade_reconciler.go:85]: not clear to me how this code works

File: operator/internal/component/registry.go (hunk 2 above)
Comment (pfmackin) [reply operator/internal/component/registry.go:47]: good question, I was actually thinking about that and don't know the answer.  We should discuss with others and decide the best order.

File: operator/controllers/upgrade_reconciler.go (hunk 4 above)
Comment (pfmackin) [reply operator/controllers/upgrade_reconciler.go:85]: ok i will add comments.

Comment (pfmackin) [reply operator/controllers/upgrade_reconciler.go]: fixed, i simplified the code also

File: operator/scripts/install/1-install-istio.sh (hunk 3 above)
Comment (pfmackin) [reply operator/scripts/install/1-install-istio.sh]: fixed

File: operator/internal/component/istio.go (hunk 1 above)
Comment (pfmackin) [reply operator/internal/component/istio.go]: fixed

File: operator/internal/component/istio_test.go [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+func TestPreUpgrade(t *testing.T) {
+	assert := assert.New(t)
+	deleteJobFunc = fakeDeletJob
+
+	chartDir, _ := filepath.Abs("./testdata")
+	err := PreUpgrade(nil, "", "ns", chartDir)
+	assert.NoError(err, "PreUpgrade returned an error")
+}
+
+func fakeDeletJob(_ clipkg.Client, jobName string, namespace string) error {
```
Comment (markxnelson) [inline operator/internal/component/istio_test.go:27]: typo

File: operator/internal/component/testdata/Chart.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# Copyright (c) 2020, Oracle and/or its affiliates.
```
Comment (markxnelson) [inline operator/internal/component/testdata/Chart.yaml:1]: update copyright notice

File: operator/internal/k8s/job.go [hunk 7]
```diff
@@ -1,7 +1,15 @@
 // Copyright (c) 2020, Oracle and/or its affiliates.
 // Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl.
 
-package internal
+package k8s
```
Comment (markxnelson) [inline operator/internal/k8s/job.go:4]: update copyright notice

Comment (markxnelson) [reply operator/internal/k8s/job.go:4]: applies to several of these files, will not repeat comment

**Review by markxnelson — CHANGES_REQUESTED** [review]

File: operator/internal/component/istio_test.go (hunk 5 above)
Comment (pfmackin) [reply operator/internal/component/istio_test.go:27]: where is the typo?

File: operator/internal/k8s/job.go (hunk 7 above)
Comment (pfmackin) [reply operator/internal/k8s/job.go:4]: sorry i missed these

File: operator/internal/component/testdata/Chart.yaml (hunk 6 above)
Comment (pfmackin) [reply operator/internal/component/testdata/Chart.yaml]: fixed

File: operator/internal/k8s/job.go (hunk 7 above)
Comment (pfmackin) [reply operator/internal/k8s/job.go:4]: all fixed

File: operator/internal/component/istio_test.go (hunk 5 above)
Comment (bobdonat) [reply operator/internal/component/istio_test.go:27]: fakeDeletJob

File: operator/scripts/install/1-install-istio.sh [hunk 8]
```diff
@@ -92,15 +92,14 @@ function install_istio()
     log "Generate istio cluster specific configuration"
     local EXTRA_HELM_ARGUMENTS=""
     if [ ${REGISTRY_SECRET_EXISTS} == "TRUE" ]; then
-      EXTRA_HELM_ARGUMENTS=" --set global.imagePullSecrets[0]=${GLOBAL_IMAGE_PULL_SECRET}"
+      EXTRA_HELM_ARGUMENTS=" --set global.imagePullSecrets[0]=${GLOBAL_IMAGE_PULL_SECRET}  --set gateways.istio-ingressgateway.type="${INGRESS_TYPE}"
```
Comment (bobdonat) [inline operator/scripts/install/1-install-istio.sh:95]: I think this additional set arg needs to be set unconditionally

Comment (pfmackin) [reply operator/scripts/install/1-install-istio.sh]: fixed

File: operator/internal/component/istio_test.go (hunk 5 above)
Comment (pfmackin) [reply operator/internal/component/istio_test.go:27]: ok

Comment (pfmackin) [reply operator/internal/component/istio_test.go]: fixed

**Review by bobdonat — APPROVED** [review]

**Review by markxnelson — APPROVED** [review]

**Pull request merged**